src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/utils/EndpointUpdateType.java

```diff
@@ -0,0 +1,179 @@
+import org.openkilda.messaging.command.haflow.HaSubFlowDto;
+import org.openkilda.model.FlowEndpoint;
+import org.openkilda.model.HaFlow;
+import org.openkilda.model.HaSubFlow;
+import org.openkilda.model.SwitchId;
+import org.openkilda.wfm.topology.flowhs.model.RequestedFlow;
+
+import org.apache.storm.shade.com.google.common.base.Objects;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public enum EndpointUpdateType {
+    NONE,
+    SOURCE,
+    DESTINATION,
+    SHARED,
+    SUBFLOWS,
+    ALL;
+
+    public boolean isPartialUpdate() {
+        return !this.equals(NONE);
+    }
+
+    /**
+     * Determines whether an update to the target HA-flow requires only endpoints update or a full update.
+     * @param originalHaFlow an original HA-flow which is to be updated
+     * @param targetHaFlow a target HA-flow from the update request
+     * @return a type of the update required
+     */
+    public static EndpointUpdateType determineUpdateType(HaFlow originalHaFlow, HaFlowRequest targetHaFlow) {
+        if (originalHaFlow == null || targetHaFlow == null) {
+            throw new IllegalArgumentException("Original and target flow object cannot be null");
+        }
+
+        boolean updateEndpointOnly = originalHaFlow.getEndpointSwitchIds().equals(getEndpointSwitchIds(targetHaFlow))
+                && Objects.equal(originalHaFlow.isAllocateProtectedPath(), targetHaFlow.isAllocateProtectedPath())
+                && originalHaFlow.getMaximumBandwidth() == targetHaFlow.getMaximumBandwidth()
+                && originalHaFlow.isIgnoreBandwidth() == targetHaFlow.isIgnoreBandwidth()
+                && originalHaFlow.isStrictBandwidth() == targetHaFlow.isStrictBandwidth()
+                && Objects.equal(originalHaFlow.getMaxLatency(), targetHaFlow.getMaxLatency())
+                && Objects.equal(originalHaFlow.getMaxLatencyTier2(), targetHaFlow.getMaxLatencyTier2())
+                && Objects.equal(originalHaFlow.getEncapsulationType(), targetHaFlow.getEncapsulationType())
+                && Objects.equal(originalHaFlow.getPathComputationStrategy(), targetHaFlow.getPathComputationStrategy())
+                && Objects.equal(originalHaFlow.getDiverseGroupId(), targetHaFlow.getDiverseFlowId());
+
+        boolean sharedEndpointChanged =
+                !originalHaFlow.getSharedEndpoint().isSwitchPortVlanEquals(targetHaFlow.getSharedEndpoint());
+
+        boolean subflowEndpointChanged = isSubflowEndpointChanged(
+                originalHaFlow.getHaSubFlows(), targetHaFlow.getSubFlows());
+
+        if (updateEndpointOnly && sharedEndpointChanged && subflowEndpointChanged) {
+            return ALL;
+        } else if (updateEndpointOnly && sharedEndpointChanged) {
+            return SHARED;
+        } else if (updateEndpointOnly && subflowEndpointChanged) {
+            return SUBFLOWS;
```

Comment:
What if we need to update rules only at the subflow1 endpoint but not on subflow2? will we update both in this case?

Reply:
I asked you this question last week that we don't have specified several things in the task, including this one. You answered that an HA flow works as a unit and we cannot see or update individual subflows.